python/servo/build_commands.py

```diff
@@ -99,11 +126,31 @@ def build(self, build_type: BuildType, jobs=None, params=None, no_package=False,
             env["RUSTFLAGS"] = env.get("RUSTFLAGS", "") + " -Zsanitizer=address"
             opts += ["-Zbuild-std"]
             kwargs["target_override"] = target_triple
-            # TODO: Investigate sanitizers in C/C++ code:
-            # env.setdefault("CFLAGS", "")
-            # env.setdefault("CXXFLAGS", "")
-            # env["CFLAGS"] += " -fsanitize=address"
-            # env["CXXFLAGS"] += " -fsanitize=address"
+
+            # Note: We want to use the same clang/LLVM version as rustc.
+            rustc_llvm_version = get_rustc_llvm_version()
+            if rustc_llvm_version is None:
+                raise RuntimeError("Unable to determine necessary clang version for ASAN support")
+            llvm_major: int = rustc_llvm_version[0]
+            target_clang = f"clang-{llvm_major}"
+            target_cxx = f"clang++-{llvm_major}"
+            if shutil.which(target_clang) is None or shutil.which(target_cxx) is None:
+                raise RuntimeError(f"--with-asan requires `{target_clang}` and `{target_cxx}` to be in PATH")
+            env.setdefault("TARGET_CC", target_clang)
+            env.setdefault("TARGET_CXX", target_cxx)
+            # TODO: We should also parse the LLVM version from the clang compiler we chose.
+            # It's unclear if the major version being the same is sufficient.
+
+            # We need to use `TARGET_CFLAGS`, since we don't want to compile host dependencies with ASAN,
+            # since that causes issues when building build-scripts / proc macros.
+            env.setdefault("TARGET_CFLAGS", "")
+            env.setdefault("TARGET_CXXFLAGS", "")
+            env["TARGET_CFLAGS"] += " -fsanitize=address"
+            env["TARGET_CXXFLAGS"] += " -fsanitize=address"
+            env["TARGET_LDFLAGS"] = "-static-libasan"
+            # Force building mozjs from source to enable ASAN with mozjs.
+            print("Info: `--with-asan` implies building `mozjs` from source. The build may take longer than usual.")
+            env["MOZJS_FROM_SOURCE"] = "1"
```

Comment:
let's set only if it's not set by env already, this way one can still set `MOZJS_FROM_SOURCE=0` and get fast compilation if it doesn't need ASAN in mozjs.

Reply:
actually that would not work right now, because:
https://github.com/servo/mozjs/blob/d1525dfaee22cc1ea9ee16c552cdeedaa9f20741/mozjs-sys/build.rs#L124
but we should fix this in mozjs